src/core/lib/surface/channel_create.cc

```diff
@@ -142,41 +144,37 @@ absl::StatusOr<grpc_channel*> CreateClientEndpointChannel(
 namespace experimental {
 
 using ::grpc_event_engine::experimental::ChannelArgsEndpointConfig;
+using ::grpc_event_engine::experimental::EndpointWrapper;
 using ::grpc_event_engine::experimental::EventEngine;
 using ::grpc_event_engine::experimental::EventEngineSupportsFdExtension;
 using ::grpc_event_engine::experimental::QueryExtension;
 
 grpc_channel* CreateChannelFromEndpoint(
-    std::unique_ptr<grpc_event_engine::experimental::EventEngine::Endpoint>
-        endpoint,
+    std::unique_ptr<EventEngine::Endpoint> endpoint,
     grpc_channel_credentials* creds, const grpc_channel_args* args) {
-  ExecCtx exec_ctx;
-  absl::StatusOr<std::string> resolved_address =
-      ResolvedAddressToURI(endpoint->GetPeerAddress());
-  auto response_generator = MakeRefCounted<FakeResolverResponseGenerator>();
-  ChannelArgs channel_args = CoreConfiguration::Get()
-                                 .channel_args_preconditioning()
-                                 .PreconditionChannelArgs(args)
-                                 .SetObject(endpoint.release());
-  if (resolved_address.ok()) {
-    channel_args =
-        channel_args.SetIfUnset(GRPC_ARG_DEFAULT_AUTHORITY, *resolved_address);
-  }
-  Resolver::Result result;
-  result.args = channel_args;
+  auto address_str = ResolvedAddressToString(endpoint->GetPeerAddress());
   // TODO(rishesh@) once https://github.com/grpc/grpc/issues/34172 is
   // resolved, we should use a different address that will be less confusing for
   // debuggability.
-  grpc_resolved_address address;
-  CHECK(grpc_parse_uri(
-      URI::Parse(resolved_address.ok() ? *resolved_address : "ipv4:0.0.0.0:0")
-          .value(),
-      &address));
+  grpc_resolved_address address =
+      CreateGRPCResolvedAddress(endpoint->GetPeerAddress());
+  auto response_generator = MakeRefCounted<FakeResolverResponseGenerator>();
+  ChannelArgs channel_args =
+      CoreConfiguration::Get()
+          .channel_args_preconditioning()
+          .PreconditionChannelArgs(args)
+          .SetObject(MakeRefCounted<EndpointWrapper>(std::move(endpoint)));
+  if (address_str.ok() && !address_str->empty()) {
+    channel_args = channel_args.SetIfUnset(
+        GRPC_ARG_DEFAULT_AUTHORITY, URI::PercentEncodeAuthority(*address_str));
+  }
+  Resolver::Result result;
+  result.args = channel_args.Remove(GRPC_ARG_SUBCHANNEL_ENDPOINT);
```

Comment:
I don't think we need to remove this arg anymore, now that we're using `RefCountedPtr<>` to store the endpoint in channel args.